src/lib.rs

```diff
@@ -51,9 +57,511 @@ mod writer;
+            MediaType::Message => "message",
+            MediaType::Image => "image",
+        }
+    }
+}
+
+impl FromStr for MediaType {
+    type Err = ParseEnumError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match s.to_ascii_lowercase().as_str() {
+            "audio" => Ok(MediaType::Audio),
+            "video" => Ok(MediaType::Video),
+            "text" => Ok(MediaType::Text),
+            "application" => Ok(MediaType::Application),
+            "message" => Ok(MediaType::Message),
+            "image" => Ok(MediaType::Image),
+            _ => Err(ParseEnumError::Invalid(s.to_string())),
+        }
+    }
+}
+
+impl Display for MediaType {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.write_str(self.as_str())
+    }
+}
+
+/// Transport Protocol for the media
+///
+/// See [RFC 8866 Section 5.14](https://datatracker.ietf.org/doc/html/rfc8866#section-5.14)
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+pub enum TransportProto {
+    /// Direct UDP
+    Udp,
+    /// RTP over UDP
+    RtpAvp,
+    /// Secure RTP over UDP
+    RtpSavp,
+    /// Secure RTP over UDP with RTCP-based feedback
+    RtpSavpf,
+}
+
+impl TransportProto {
+    pub fn as_str(&self) -> &'static str {
+        match self {
+            // The strings are case-insensitive, but the spec (RFC 8866) uses lower-case of the "udp" protocol
+            // and upper-case for the others so keeping it the same
+            TransportProto::Udp => "udp",
+            TransportProto::RtpAvp => "RTP/AVP",
+            TransportProto::RtpSavp => "RTP/SAVP",
+            TransportProto::RtpSavpf => "RTP/SAVPF",
+        }
+    }
+}
+
+impl FromStr for TransportProto {
+    type Err = ParseEnumError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        // The strings are case-insensitive, but the spec (RFC 8866) uses lower-case of the "udp" protocol
+        // and upper-case for the others so keeping it the same
+        if "udp".eq_ignore_ascii_case(s) {
+            Ok(TransportProto::Udp)
+        } else if "RTP/AVP".eq_ignore_ascii_case(s) {
+            Ok(TransportProto::RtpAvp)
+        } else if "RTP/SAVP".eq_ignore_ascii_case(s) {
+            Ok(TransportProto::RtpSavp)
+        } else if "RTP/SAVPF".eq_ignore_ascii_case(s) {
+            Ok(TransportProto::RtpSavpf)
+        } else {
+            Err(ParseEnumError::Invalid(s.to_string()))
+        }
+    }
+}
+
+impl Display for TransportProto {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.write_str(self.as_str())
+    }
+}
+
+/// RtpMap Attribute
+///
+/// See [RFC 8866 Section 6.6](https://datatracker.ietf.org/doc/html/rfc8866#section-6.6) for more details
+#[derive(Debug, Clone, PartialEq)]
+pub struct RtpMap {
+    /// Payload type, a numerical value between 0 and 127
+    pub pt: u8,
+    /// Name of the encoding
+    // TODO: is it useful to have an enum for all known encoding?
```

Comment:
Potentially useful, just like for the registered payload types. But in this struct here, a string / u8 is the correct thing. You could have functions on the struct that try to handle pt / encoding with those enums though

Reply:
https://gitlab.freedesktop.org/gstreamer/gst-plugins-rs/-/blob/44a632811e6718909409b5c229508072a1447e62/net/rtsp/src/rtspsrc/sdp.rs#L215 basically having `guess_rtpmap_from_pt()` and `guess_rtpmap_from_encoding_name()` here in some form seems useful.

@nirbheek Do you remember why your code allows rtpmap without clock-rate? The SDP RFC does not allow that

Reply:
There are payload types for which clock-rate is static, and tools like ffmpeg won't add the clock-rate in those cases. For example pt=10/11 (L16 2ch/1ch), see table in https://en.wikipedia.org/wiki/RTP_payload_formats#Audio_and_video_payload_types.